### PR TITLE
Handle explicit segmentation keys

### PR DESCRIPTION
**Pull request by richard-rogers**
## Description

Allow `segment_key_values` in addition to segmented `DatasetSchema`

- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

File: python/whylogs/api/logger/logger.py [hunk 1]
```diff
@@ -114,16 +114,15 @@ def log(
         # If segments are defined use segment_processing to return a SegmentedResultSet
         if active_schema and active_schema.segments:
             if segment_key_values:
-                raise ValueError(
-                    f"using explicit `segment_key_values` {segment_key_values} is not compatible "
-                    f"with segmentation also defined in the DatasetSchema: {active_schema.segments}"
-                )
+                pass  # TODO: maybe check for duplicate keys
```
Comment (richard-rogers) [inline python/whylogs/api/logger/logger.py]: Better to do it in `_log_segment()`?

File: python/whylogs/api/logger/segment_processing.py [hunk 2]
```diff
@@ -129,15 +131,19 @@ def _log_segment(
...
     segments: Dict[Segment, Any] = {}
     pandas, row = _pandas_or_dict(obj, pandas, row)
     if partition.filter:
+        # TODO: segment_key_values don't apply to filter since they're explicitely added
         pandas, row = _filter_inputs(partition.filter, pandas, row)
     if partition.simple:
         columns = partition.mapper.col_names if partition.mapper else None
         if columns:
-            _process_simple_partition(partition.id, schema, segments, columns, pandas, row, segment_cache)
+            _process_simple_partition(
```
Comment (richard-rogers) [inline python/whylogs/api/logger/segment_processing.py:145]: Should we allow empty columns if `segment_key_values` are specified?

File: python/whylogs/api/logger/logger.py (hunk 1 above)
Comment (jamie256) [reply python/whylogs/api/logger/logger.py]: flexible on where we check, it does seem better in `_log_segment`

File: python/whylogs/api/logger/segment_processing.py (hunk 2 above)
Comment (jamie256) [reply python/whylogs/api/logger/segment_processing.py:145]: Its an edge case, not sure I understand how we get an empty column set with a simple segmentation partition, but if we do have an explicit segment via `segment_key_values` it would be good to handle this like the explicit segmentation route.

File: python/whylogs/api/logger/__init__.py [hunk 3]
```diff
@@ -165,6 +165,7 @@ def log_classification_metrics(
     schema: Optional[DatasetSchema] = None,
     log_full_data: bool = False,
     dataset_timestamp: Optional[datetime] = None,
+    # TODO: should we add segment_key_values here?
```
Comment (jamie256) [inline python/whylogs/api/logger/__init__.py]: Yes, but can be in a separate PR.

File: python/whylogs/api/logger/segment_processing.py [hunk 4]
```diff
@@ -129,15 +133,19 @@ def _log_segment(
     pandas: Optional[pd.DataFrame] = None,
     row: Optional[Mapping[str, Any]] = None,
     segment_cache: Optional[SegmentCache] = None,
+    segment_key_values: Optional[Dict[str, str]] = None,
 ) -> Dict[Segment, Any]:
     segments: Dict[Segment, Any] = {}
     pandas, row = _pandas_or_dict(obj, pandas, row)
     if partition.filter:
+        # TODO: segment_key_values don't apply to filter since they're explicitely added
```
Comment (richard-rogers) [inline python/whylogs/api/logger/segment_processing.py]: Should verify this assumption

File: python/whylogs/api/logger/__init__.py (hunk 3 above)
Comment (richard-rogers) [inline python/whylogs/api/logger/__init__.py]:
```suggestion
```

File: python/whylogs/api/logger/segment_processing.py (hunk 2 above)
Comment (richard-rogers) [reply python/whylogs/api/logger/segment_processing.py:145]: I knocked up some code to handle the case, but I'm inclined to think of this as a user error -- use an unsegmented schema and `segment_key_values` for this case. Maybe throw or log an error here?

Comment (jamie256) [reply python/whylogs/api/logger/segment_processing.py:145]: I'm ok with logging an error here, hinting on the correct usage pattern.

File: python/whylogs/api/logger/segment_processing.py (hunk 4 above)
Comment (jamie256) [reply python/whylogs/api/logger/segment_processing.py]: The comment looks correct, can remove the TODO part of it

**Review by jamie256 — DISMISSED** [review]
Looks good. Maybe cleanup some of the comments but the changes look right.

**Review by jamie256 — APPROVED** [review]
LGTM!
